parser_library/src/debugging/debugger.cpp

```diff
@@ -14,271 +14,431 @@
-{
-    opencode_source_path_ = open_code->get_file_name();
-    auto b = std::bind(&debugger::debug_start, this, std::placeholders::_1, std::placeholders::_2);
-    stop_on_next_stmt_ = stop_on_entry;
+    // these are used in conditional variable to stop execution
+    // of analyzer and wait for user input
+    std::condition_variable con_var;
+    std::atomic<bool> continue_ = false;
 
-    thread_ = std::make_unique<std::thread>(b, open_code, &provider);
-}
+    bool debug_ended_ = false;
 
-void debugger::statement(range stmt_range)
-{
-    if (disconnected_)
-        return;
+    // Specifies whether the debugger stops on the next statement call.
+    std::atomic<bool> stop_on_next_stmt_ = false;
+    std::atomic<bool> step_over_ = false;
+    size_t step_over_depth_ = 0;
+
+    // Range of statement that is about to be processed by analyzer.
+    range next_stmt_range_;
+
+    // True, if disconnect request was received
+    bool disconnected_ = false;
+
+    // Cancellation token for parsing
+    std::atomic<bool> cancel_ = false;
 
-    bool breakpoint_hit = false;
-    auto bps = cfg_.get_breakpoints(ctx_->processing_stack().back().proc_location.file);
-    for (auto& bp : bps.points)
+    // Provides a way to inform outer world about debugger events
+    debug_event_consumer* event_ = nullptr;
+
+    // Debugging information retrieval
+    context::hlasm_context* ctx_ = nullptr;
+    std::string opencode_source_path_;
+    std::vector<stack_frame> stack_frames_;
+    std::vector<scope> scopes_;
+
+    std::unordered_map<size_t, variable_store> variables_;
+    size_t next_var_ref_ = 1;
+    context::processing_stack_t proc_stack_;
+
+    std::unordered_map<std::string, std::vector<breakpoint>> breakpoints_;
+
+    size_t add_variable(std::vector<variable_ptr> vars)
     {
-        if (bp.line >= stmt_range.start.line && bp.line <= stmt_range.end.line)
-            breakpoint_hit = true;
+        variables_[next_var_ref_].variables = std::move(vars);
+        return next_var_ref_++;
     }
 
+public:
+    impl() = default;
 
-    // breakpoint check
-    if (stop_on_next_stmt_ || breakpoint_hit || (step_over_ && ctx_->processing_stack().size() <= step_over_depth_))
+    void launch(
+        std::string_view source, workspaces::workspace& workspace, bool stop_on_entry, parse_lib_provider* lib_provider)
     {
-        variables_.clear();
-        stack_frames_.clear();
-        scopes_.clear();
-        proc_stack_ = ctx_->processing_stack();
-        variable_mtx_.unlock();
+        // TODO: check if already running???
+        auto open_code = workspace.get_processor_file(std::string(source));
+        opencode_source_path_ = open_code->get_file_name();
+        stop_on_next_stmt_ = stop_on_entry;
+
+        thread_ = std::thread([this, open_code = std::move(open_code), &workspace, lib_provider]() {
+            std::lock_guard<std::mutex> guard(variable_mtx_); // this is odd
```

Comment:
The idea is, that `variable_mtx_` should be locked always when the analyzer is running (and therefore context is being changed) the mutex is only unlocked once the control flow of analyzer is within this class, waiting in the `statement` method. Only then can we safely access the context and see the variables inside it.
```suggestion
            std::lock_guard<std::mutex> guard(variable_mtx_); // Lock the mutex while analyzer is running, unlock once it is stopped and waiting in the statement method
```

Reply:
Yeah, I eventually got that, nevertheless, we use the lock_guard here, then secretly unlock/re-lock the lock later in a different place.